indico/web/client/js/react/components/DatePicker.jsx

```diff
@@ -0,0 +1,99 @@
+// This file is part of Indico.
+// Copyright (C) 2002 - 2024 CERN
+//
+// Indico is free software; you can redistribute it and/or
+// modify it under the terms of the MIT License; see the
+// LICENSE file for more details.
+
+import moment from 'moment';
+import PropTypes from 'prop-types';
+import React from 'react';
+
+import {Param, Translate} from 'indico/react/i18n';
+import {formatDate} from 'indico/utils/date_format';
+import {fromISOLocalDate} from 'indico/utils/date_parser';
+
+import 'indico/custom_elements/ind_date_picker';
+
+export default function DatePicker({
+  onChange,
+  value,
+  format = moment.localeData().longDateFormat('L'),
+  ...inputProps
+}) {
+  function handleDateChange(ev) {
+    const {date} = ev.target.closest('ind-date-picker');
+    // en-CA uses the ISO format date (YYYY-MM-DD) but gives it in local time zone.
+    // Do not use toISOString() for this because it may result in incorrect date due
+    // to time zone differences.
+    onChange(date?.toLocaleDateString('en-CA'));
+  }
+
+  const formattedValue = formatDate(format, fromISOLocalDate(value));
+
+  return (
+    <ind-date-picker>
+      <input
+        type="text"
+        onChange={handleDateChange}
+        defaultValue={formattedValue}
+        {...inputProps}
+      />
+      <button type="button" disabled={inputProps.disabled}>
+        <Translate as="span">Open a calendar</Translate>
+      </button>
+      <ind-calendar>
+        <dialog>
+          <div className="controls">
+            <button type="button" value="previous-year">
+              <Translate as="span">Previous year</Translate>
+            </button>
+            <button type="button" value="previous-month">
+              <Translate as="span">Previous month</Translate>
+            </button>
+
+            <div className="month-year">
+              <select />
+              <input type="number" required />
+            </div>
+
+            <button type="button" value="next-month">
+              <Translate as="span">Next month</Translate>
+            </button>
+            <button type="button" value="next-year">
+              <Translate as="span">Next year</Translate>
+            </button>
+          </div>
+
+          <div className="weekdays">
+            <span />
+            <span />
+            <span />
+            <span />
+            <span />
+            <span />
+            <span />
+          </div>
+          <div role="listbox" />
+        </dialog>
+      </ind-calendar>
+
+      <span className="date-format">
+        <Translate>
+          Date format: <Param name="format" value={format} />
```

Comment:
What do you think about showing this as a `placeholder` in the input field? I think maybe this whole date-format block below could go away altogether then.

Reply:
Placeholder isn't used by screen readers, so the date format block must be there. That's why I omitted the placeholder.

Reply:
Isn't there a way to hide it from regular users but still expose it to screen readers etc? Maybe using negative margin or whatever hack is often used for things like those skip links at the beginning of the page?

I think especially in places where we use the locale default and thus don't want to show the date format at all, the placeholder would look nicer.

Actually it already looks a bit weird in the registration form when there's also a field description:
![image](https://github.com/indico/indico/assets/179599/e33c2176-e538-4c8e-88bd-f2a042222a2d)

Reply:
I will make that change. However, I will also note that from the UX perspective, placeholder is generally an anti-pattern. Not a great way of conveying anything. Suppose you start typing 2012 and you get distracted and you get back to the field. You would no longer know what the format should be without removing the input. Not great. If the information matters at all, it should be present outside the field. If it doesn't matter, it should not be present at all.

Reply:
`const formatDescription = this.querySelector('.date-format');` now fails because styleName applies CSS module name mangling

Reply:
Fixed